libsolidity/parsing/Scanner.cpp

```diff
@@ -780,19 +780,19 @@ Token::Value Scanner::scanNumber(char _charSeen)
 		{
 			addLiteralCharAndAdvance();
 			// either 0, 0exxx, 0Exxx, 0.xxx or a hex number
-			if (m_char == 'x' || m_char == 'X')
+			if (m_char == 'x')
 			{
 				// hex number
 				kind = HEX;
 				addLiteralCharAndAdvance();
 				if (!isHexDigit(m_char))
-					return Token::Illegal; // we must have at least one hex digit after 'x'/'X'
+					return Token::Illegal; // we must have at least one hex digit after 'x'
 
 				while (isHexDigit(m_char) || m_char == '_') // We keep the underscores for later validation
 					addLiteralCharAndAdvance();
 			}
-			else if (isDecimalDigit(m_char))
-				// We do not allow octal numbers
+			else if (isDecimalDigit(m_char) || m_char == 'X')
+				// We do not allow octal numbers and 'X' in hex numbers
```

Comment:
I don't think the special check for `X` is needed here. It should fall through the next step if wnad fail there.

Reply:
I misread this part. I'll remove the special check for `X` here.